File: python/graphs/1514-path-with-maximum-probability.py
import heapq

from collections import defaultdict
from typing import List


def maxProbability(n: int, edges: List[List[int]], succProb: List[float], start: int, end: int) -> float:
    graph = defaultdict(list)

    for i in range(len(edges)):
        graph[edges[i][0]].append([edges[i][1], succProb[i]])
        graph[edges[i][1]].append([edges[i][0], succProb[i]])
        
    probs = [[-1, start]]
    
    heapq.heapify(probs)
    visited = set()
    
    while probs:
        cur_prob, cur = heapq.heappop(probs)
        visited.add(cur)
        
        if cur == end:
            return -1 * cur_prob
        for neighbor in graph[cur]:
            if neighbor[0] not in visited:
                heapq.heappush(probs, [cur_prob * neighbor[1], neighbor[0]])

    return 0

File: cpp/graphs/1514-path-with-maximum-probability.cpp
#include <queue>
#include <set>
#include <utility>
#include <vector>

using namespace std;

double maxProbability(int n, vector<vector<int>>& edges, vector<double>& succProb, int start, int end) {
    vector<vector<pair<int, double>>> adj_list (n);
    
    for (int i = 0; i < edges.size(); i++) {
        adj_list[edges[i][0]].push_back({edges[i][1], succProb[i]});
        adj_list[edges[i][1]].push_back({edges[i][0], succProb[i]});
    }
    
    priority_queue<pair<double, int>> q;
    set<int> visited;
    
    q.push({1, start});
    
    while (!q.empty()) {
        auto cur = q.top();
        q.pop();
        
        if (cur.second == end) return cur.first;
        visited.insert(cur.second);
        
        for (const auto neighbor : adj_list[cur.second]) {
            if (visited.find(neighbor.first) == visited.end()) {
                q.push({cur.first * neighbor.second, neighbor.first});
            }
        }
    }
    
    return 0;
}